File: 程序员面试金典/8.9括号/generateParenthesis.py
class Solution:
    def __init__(self):
        self.ans = []
    
    def generateParenthesis(self, n: int) -> List[str]:
        self.generate(n, 0, [])
        return self.ans
    
    def generate(self, n_left, n_unclosed, res):
        if n_left == 0:
            self.ans.append(''.join(res) + ')' * n_unclosed)
        else:
            res.append('(')
            self.generate(n_left - 1, n_unclosed + 1, res)
            res.pop()
            if n_unclosed > 0:
                res.append(')')
                self.generate(n_left, n_unclosed - 1, res)
                res.pop()

File: 程序员面试金典/8.9/generateParenthesis.java
class Solution {
    private List<String> res = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        char[] a = new char[2 * n];
        generate(a, 0, n, n);
        return res;
    }

    private void generate(char[] a, int index, int left, int right) {
        if (left < 0 || right < left)
            return;
        if (left == 0 && right == 0) {
            res.add(new String(a));
            return;
        }
        a[index] = '(';
        generate(a, index + 1, left - 1, right);
        a[index] = ')';
        generate(a, index + 1, left, right - 1);
    }

}
